File: 2024/day3.cpp
#include <iostream>
#include <string>

#include <re2/re2.h>
#include <re2/stringpiece.h>
#include "../util.h"

using ulli = unsigned long long int;

int main(int, char**){
    std::string inp = util::getInput(2024, 3);

    // part 1
    RE2 mulPatt("mul\\((\\d+),(\\d+)\\)");
    int m1, m2;
    ulli res = 0;
    {
        re2::StringPiece inp_view(inp);
        while (RE2::FindAndConsume(&inp_view, mulPatt, &m1, &m2)){
            res += m1 * m2;
        }
    }
    std::cout << res << std::endl;
    
    // part 2
    res = 0;
    size_t ptr = 0, loc;
    while (ptr != std::string::npos){
        loc = inp.find("don't()", ptr);
        if (loc == std::string::npos){
            loc = inp.size();
        }
        {
            re2::StringPiece substr_view = re2::StringPiece(inp).substr(ptr, loc - ptr);
            while (RE2::FindAndConsume(&substr_view, mulPatt, &m1, &m2)){
                res += m1 * m2;
            }
        }
        ptr = inp.find("do()", loc);
    }
    std::cout << res << std::endl;

    return 0;
}

File: 2024/day3.py
import AOCInit
import util

import re

if __name__ != '__main__':
    exit()

inp = util.getInput(d=3, y=2024)

# part 1
mulPatt = re.compile(r'mul\((\d+),(\d+)\)')
print(sum(int(a) * int(b) for a, b in mulPatt.findall(inp)))

# part 2
doPatt = re.compile(r'do\(\)')
dontPatt = re.compile(r"don't\(\)")
n = len(inp)
ptr = 0
res = 0
while ptr < n:
    m = dontPatt.search(inp[ptr:])
    loc = n
    if m is not None:
        loc = ptr + m.start()
    res += sum(int(a) * int(b) for a, b in mulPatt.findall(inp[ptr:loc]))
    m = doPatt.search(inp[loc:])
    if m is None:
        break
    ptr = loc + m.start()
print(res)

